### DISCUSSION TITLE
How do I rotate beyond the 'limit' vertically?

### DISCUSSION THREAD

**Discussion by jsegeren** [google/model-viewer#3314] [Q&A]
Hi, I want to rotate the model beyond the standard 'vertical' limit. 
Currently it looks like phi = 2.75 rad is the limit. I want to effectively rotate more than "2.75 rad".
Actually, I want to be able to rotate in both directions (horizontal aka theta and vertical aka phi) "infinitely".
Any suggestions on how to do that?
Thanks!

**Comment by elalish** [comment]
We don't allow you to rotate the model upside-down, since that can be confusing to users. However, you can set the min and max camera-orbit to get full 0-180: https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-maxCameraOrbit

If you really need the model upside down, you can use the `orientation` attribute.

**Reply by jsegeren** [reply]
Thank you @elalish !
Our use case is a bit different. It's very important for the user experience to be able to rotate the model continuously in any direction. 
I looked at 'orientation' - doesn't appear a viable solution, since the update framing call would be apparent and jarring.

It'd be extremely helpful - and likely useful to the community - if you supported 'infinite' for camera-orbit (theta, phi).

**Reply by elalish** [reply]
We'll consider it if you can show us your use case. Do you have an example, or at least a description?

**Reply by jsegeren** [reply]
It's a stealth project so I can't fully discuss here. (Happy to message you privately... ) 
Basically the interaction UX lets the user "play" with the model - e.g. swipe to "spin" it (gives it a certain speed and momentum). It's weird for the model to just 'hit a wall' when the orbit hits the limit, vs continuing to spin in that direction with momentum. 
Thanks for considering.

**Reply by elalish** [reply]
Well, go ahead and send us a PR to review. It might be pretty simple. Take a look at SmoothControls.ts.

**Reply by jsegeren** [reply]
Hey @elalish - thanks, tried making a few minor adjustments, but it didn't work. 
Looks like the change could require a bit more knowledge of the codebase - a bit beyond what I can take on in very short-term.
Would your team possibly consider it as a DCR or feature request?

**Reply by elalish** [reply]
You're the first to ask for it, so it's a fairly low priority for us, but yes, it's been noted. It will happen much sooner with your own PR than waiting for us.

**Reply by daniel-packard** [reply]
not the first :) 

https://github.com/google/model-viewer/discussions/3219

I'm happy to dig in and PR if you're open to the change

**Reply by elalish** [reply]
Absolutely, always open to a PR! Thanks.

**Reply by Shalwaban** [reply]
sounds like the trackball controls
https://github.com/mrdoob/three.js/blob/master/examples/jsm/controls/TrackballControls.js

**Comment by jsegeren** [comment]
Fair enough. Thanks!

On Mon, Apr 4, 2022, 10:16 AM Emmett Lalish ***@***.***>
wrote:

> You're the first to ask for it, so it's a fairly low priority for us, but
> yes, it's been noted. It will happen much sooner with your own PR than
> waiting for us.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/google/model-viewer/discussions/3314#discussioncomment-2502204>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAUT2UWCPHUPOEEKKMZRV53VDMPYFANCNFSM5RZS5V7Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

-- 



The content of this message is the proprietary and confidential property 
of Fluid Inc, and should be treated as such. If you are not the intended 
recipient and have received this message in error, please delete this 
message from your computer system and notify me immediately by reply 
e-mail. Any unauthorized use or distribution of the content of this message 
is prohibited.

**Comment by protokoll-studio** [comment]
Hello. Any news about the possibilty to rotate at 360° in any direction? That would be really nice

**Reply by elalish** [reply]
Can you give a link to a 3D UX (live or video) that you like? We intentionally keep model UP (+Y) always pointing to the top of the screen so users don't get lost. You can always turn off camera-controls and make your own pointer event listeners. You might be interested in my blog article on 3D UX and the design philosophy I've used: https://elalish.blogspot.com/2022/04/3d-interaction.html

**Reply by dsvatunek** [reply]
I would like to use it for chemical models and basically every chemical model viewer (e.g. 3Dmol.js) allows you to rotate in every axis indefinitely. There's no up or down in these models so it makes sense to be able to rotate in any way.

**Reply by elalish** [reply]
You're welcome to grab the trackball snippet from my blog post above, but our API won't allow going past the poles because it's not as simple as just expanding the allowable values.

**Reply by protokoll-studio** [reply]
Hey ! I tried your trackball solution. It worked but now I need to add the possibility to zoom. But nothing is impossible ;)
I would be a nice improvement if we can have a native solution.

**Reply by elalish** [reply]
Thanks, but we have to be opinionated about our UX to keep our quality high. As a user, I would still strongly encourage you to pick a sensible up direction, as it makes it easier to orient yourself in 3D. My blog post has more reasons.
